### ISSUE TITLE
Anchors for headings

### ISSUE THREAD

**Issue by neovov** [uranusjr/macdown#255]
Hi,

I noticed that the generated anchors on headings are toc_xx.
Is it possible to implement something more relevant like GitHub? This is especially useful if you want to provide inner links or insert a custom table of content.

Instead of:
```html
<h1 id="toc_1">Welcome to my project</h1>
```

Generate this:
```html
<h1 id="welcome-to-my-project">Welcome to my project</h1>
```

Thanks for the good work :+1: :beers: !

**Comment by uranusjr** [comment]
I don’t think it would be too hard, but one question: How does GitHub handle things if there are multiple headers with the same name? Something like:

    # Foo
    ## Bar
    # Bar
    ## Bar

The three headers can’t all have ID `bar`. But what should this produce? If GItHub has a good solution for this, I might be able to implement the rest as soon as this weekend. :beer:

**Comment by neovov** [comment]
It will indent the other occurrences:

```html
<h1 id="foo">Foo</h1>
<h2 id="bar">Bar</h1>
<h1 id="bar-1">Bar</h1>
<h2 id="bar-2">Bar</h1>
```

It doesn't seems too complicated to implement.
Another enhancement would let the possibility to create our own HTML templates but it might be a bit complex.

:heart:

**Comment by uranusjr** [comment]
I played a bit with GitHub’s ID-generating algorithm, and to be honest it is pretty quirky, to say at least, when it comes to non-ASCII characters, and may even generate invalid HTML if you feed it with very funky inputs. I’ve contacted GitHub to clarify on some potential bugs, and will probably wait until at least they respond to make the next move. The implementation is almost done, but I would really hate to leave things undone. This will probably take some time.

**Comment by neovov** [comment]
Thanks for the good work and the heads up ! :+1:

**Comment by nchammas** [comment]
On a related note (and perhaps this should be a separate issue), it would be great if you could click on an anchor link that references a section in the currently open document, and MacDown would jump to that section. That would match how intra-page anchor links work on the web.

**Comment by Jmuccigr** [comment]
How about matching what pandoc does? 

http://pandoc.org/README.html#header-identifiers-in-html-latex-and-context

**Comment by Mark-L6n** [comment]
+1 For better naming than `toc_xx`—I use internal links quite a bit.
1) Auto-generating name with dashes:  there are probably some libraries for automatically creating web path names that could be useful, such as those with content management systems.  See for ex. [Drupal Pathauto](https://www.drupal.org/documentation/modules/pathauto)
2) As for duplicates, a quick hack would be not to worry about them.  If there are multiple identical IDs, my browser (Chrome) simply goes to the 1st one.  It'd be up to the author to not write duplicates.

**Comment by Jmuccigr** [comment]
For the record, here's how pandoc does it, taken from the link I posted above:

A header without an explicitly specified identifier will be automatically assigned a unique identifier based on the header text. To derive the identifier from the header text,

* Remove all formatting, links, etc.
* Remove all footnotes.
* Remove all punctuation, except underscores, hyphens, and periods.
* Replace all spaces and newlines with hyphens.
* Convert all alphabetic characters to lowercase.
* Remove everything up to the first letter (identifiers may not begin with a number or punctuation mark).
* If nothing is left after this, use the identifier section.

Thus, for example,

|Header|Identifier|
|---|---|
|Header identifiers in HTML	|header-identifiers-in-html|
|Dogs?–in my house?	|dogs--in-my-house|
|HTML, S5, or RTF?	|html-s5-or-rtf|
|3. Applications	|applications|
|33	|section|

These rules should, in most cases, allow one to determine the identifier from the header text. The exception is when several headers have the same text; in this case, the first will get an identifier as described above; the second will get the same identifier with -1 appended; the third with -2; and so on.

**Comment by franklinyu** [comment]
@Jmuccigr There is an [en dash](http://unicode-table.com/en/2013/) in `Dogs?–in my house?`. I guess you didn't mean it. 😄

**Comment by Jmuccigr** [comment]
Just look at the original, please. I was copying and pasting.

**Comment by franklinyu** [comment]
@Jmuccigr Yeah I saw that. The anchor for the is updated, though; it is now [Extension: `auto_identifiers`](http://pandoc.org/MANUAL.html#extension-auto_identifiers).

**Comment by Jmuccigr** [comment]
Yes, well it was **over a year ago**.

Was there something you wanted to add to this discussion? :-)

**Comment by franklinyu** [comment]
No, that's it. I'm still researching whether this feature belongs to the Hoedown engine.

**Comment by xguerin** [comment]
Any update on this issue ?

**Comment by utdrmac** [comment]
Any updates on this issue? Having anchors would be great.

**Comment by johndavidmiller** [comment]
Yes, according to https://stackoverflow.com/questions/2822089/how-to-link-to-part-of-the-same-document-in-markdown the "standard" Markdown syntax is 

    [some name](#my-linked-header)

 to link to 

    ### My Linked Header

Note that: 1) there is only one # in the reference, regardless of Header depth, 2) references are lower-case, and 3) multiple words are separated by dashes.

MacDown does so many other things awesomely; not implemented intra-document links is a glaring omission.

**Comment by morningmorningwind** [comment]
I'm looking for an answer. Is this problem solved? Thanks.

**Comment by franklinyu** [comment]
@utdrmac I would not call the project a dead one, given that last commit in master branch was only 16 days ago, and we had testing release this April. I assume @uranusjr just want to keep stable release really stable.

**Comment by alexisglr** [comment]
+1 This feature would be very usefull !

**Comment by calypso15** [comment]
+1. Has there been an update on this feature?

**Comment by utdrmac** [comment]
Doesn't look like this project has had a serious update since this time last year. Is this app dead?

**Comment by Mark-L6n** [comment]
FYI, my hack to get around this has been to insert a `div` with `id`, like:

```
* [Internal Links](#int_links)
…
…
<div id="int_links"></div>
## Internal Links
```

**Comment by mdaeron** [comment]
+1 for this feature. Easy anchor links to headers would be so useful! Is there any chance to eventually implement this?

**Comment by jcheek** [comment]
+1 for this, still, even if it's just Microsoft's simple model at https://learn.microsoft.com/en-us/azure/devops/project/wiki/markdown-guidance?view=azure-devops#anchor-links .

Creating anchors named toc-XX is a non-starter, since the XX will change whenever a heading is added or removed.

<blockquote><img src="https://learn.microsoft.com/en-us/media/open-graph-image.png" width="48" align="right"><div><strong><a href="https://learn.microsoft.com/en-us/azure/devops/project/wiki/markdown-guidance?view=azure-devops">Markdown syntax for files, widgets, wikis - Azure DevOps</a></strong></div><div>Discover how to utilize Markdown to incorporate tables, mathematical symbols, and share information effectively in pull requests, README files, dashboards, and wikis in Azure DevOps.</div></blockquote>

**Comment by barrypress** [comment]
Please, if you do nothing else to macdown in 2024, please implement this internal link "standard" !!!

**Comment by dmland** [comment]
+1 – Even accepting that there are multiple "flavors" of markdown, this is basically table stakes, and it's been a full decade that people have asked for it.
